File: solutions/elixir/list-ops/1/lib/list_ops.ex
defmodule ListOps do
  # Please don't use any external modules (especially List or Enum) in your
  # implementation. The point of this exercise is to create these basic
  # functions yourself. You may use basic Kernel functions (like `Kernel.+/2`
  # for adding numbers), but please do not use Kernel functions for Lists like
  # `++`, `--`, `hd`, `tl`, `in`, and `length`.

  @spec count(list) :: non_neg_integer
  def count(list, acc\\0)
  def count([], acc), do: acc
  def count([_head|tail], acc), do: count(tail,acc+1)

  @spec reverse(list) :: list
  def reverse(l,acc\\[])
  def reverse([], acc), do: acc
  def reverse([head|tail], acc), do: reverse(tail,[head|acc])

  @spec map(list, (any -> any)) :: list
  def map([], _f), do: []
  def map([head|tail], f), do: [f.(head)| map(tail, f)]

  @spec filter(list, (any -> as_boolean(term))) :: list
  def filter([], _f), do: []
  def filter([head|tail], f) do
    if not f.(head) do
      filter(tail, f)
    else
      [head|filter(tail, f)]
    end
  end

  @type acc :: any
  @spec reduce(list, acc, (any, acc -> acc)) :: acc
  def reduce([], acc, _f), do: acc
  def reduce([head|tail], acc, f), do: reduce(tail, f.(head, acc), f)

  @spec append(list, list) :: list
  # @spec append(any, any) :: nil
  def append([],[]), do: []
  def append([], b), do: b
  def append([head|tail], b), do: [head|append(tail, b)]

  @spec concat([[any]]) :: [any]
  def concat(ll) do
    # reduce with an append funciton
    reverse(ll)|>
    reduce( [], &append/2)
  end
end

File: solutions/erlang/list-ops/1/src/list_ops.erl
-module(list_ops).

-export([append/2, concat/1, filter/2, length/1, map/2, foldl/3, foldr/3, reverse/1]).

append(List1, List2) ->
    append(List1, List2, []).

append([], [], Acc) ->
    reverse(Acc);
append([], [Head | Tail], Acc) ->
    append([], Tail, [Head | Acc]);
append([Head | Tail], List2, Acc) ->
    append(Tail, List2, [Head | Acc]).

concat(List) ->
    concat(List, []).

concat([], Acc) ->
    Acc;
concat([Head | Tail], Acc) ->
    concat(Tail, append(Acc, Head)).

filter(Function, List) ->
    filter(Function, List, []).

filter(_Function, [], Acc) ->
    reverse(Acc);
filter(Function, [Head | Tail], Acc) ->
    case Function(Head) of
        true ->
            filter(Function, Tail, [Head | Acc]);
        false ->
            filter(Function, Tail, Acc)
    end.

length(List) ->
    length(List, 0).

length([], Sum) ->
    Sum;
length([_Head | Tail], Sum) ->
    length(Tail, Sum + 1).

map(Function, List) ->
    map(Function, List, []).

map(_Function, [], Acc) ->
    reverse(Acc);
map(Function, [Head | Tail], Acc) ->
    map(Function, Tail, [Function(Head) | Acc]).

foldl(_Function, Start, []) ->
    Start;
foldl(Function, Start, [Head | Tail]) ->
    foldl(Function, Function(Head, Start), Tail).

foldr(Function, Start, List) ->
    foldl(Function, Start, reverse(List)).

reverse(List) ->
    reverse(List, []).

reverse([], Acc) ->
    Acc;
reverse([Head | Tail], Acc) ->
    reverse(Tail, [Head | Acc]).
